File: test/category_tree_test.dart
import 'dart:async';

import 'package:flutter_riverpod/flutter_riverpod.dart';
import 'package:flutter_test/flutter_test.dart';
import 'package:sudan_it_marketplace/features/categories/data/models/category_model.dart';
import 'package:sudan_it_marketplace/features/categories/domain/category_tree.dart';
import 'package:sudan_it_marketplace/features/categories/domain/entities/category.dart';
import 'package:sudan_it_marketplace/features/categories/presentation/category_providers.dart';

final _created = DateTime(2026);

/// A category; the ancestor chain is worked out from [parentId] by [tree].
Category _c(
  String id,
  String? parentId, {
  String ar = '',
  String? en,
  int? sortOrder,
  bool isActive = true,
  bool deletionPending = false,
}) => Category(
  id: id,
  name: en ?? id,
  nameAr: ar,
  nameEn: en ?? id,
  description: '',
  iconName: '',
  isActive: isActive,
  createdAt: _created,
  sortOrder: sortOrder,
  parentId: parentId,
  deletionPending: deletionPending,
);

void main() {
  group('CategoryTree', () {
    final sample = [
      _c('net', null, ar: 'شبكات', en: 'Networking', sortOrder: 0),
      _c('rou', 'net', ar: 'راوترات', en: 'Routers', sortOrder: 0),
      _c('wifi', 'rou', ar: 'واي فاي', en: 'Wi-Fi 6'),
      _c('swi', 'net', ar: 'سويتشات', en: 'Switches', sortOrder: 1),
      _c('lap', null, ar: 'لابتوبات', en: 'Laptops', sortOrder: 1),
    ];

    test('lists children in the order Platform Admin set, at every level', () {
      final tree = CategoryTree(sample);
      expect(tree.childrenOf(null).map((c) => c.id), ['net', 'lap']);
      expect(tree.childrenOf('net').map((c) => c.id), ['rou', 'swi']);
      expect(tree.childrenOf('rou').map((c) => c.id), ['wifi']);
      expect(tree.childrenOf('wifi'), isEmpty);
      expect(tree.hasChildren('net'), isTrue);
      expect(tree.hasChildren('wifi'), isFalse);
    });

    test('gives the full path in the language shown', () {
      final tree = CategoryTree(sample);
      expect(tree.pathOf('wifi').map((c) => c.id), ['net', 'rou', 'wifi']);
      expect(tree.pathLabel('wifi', 'en'), 'Networking › Routers › Wi-Fi 6');
      expect(tree.pathLabel('wifi', 'ar'), 'شبكات › راوترات › واي فاي');
      expect(tree.pathLabel('lap', 'en', separator: ' / '), 'Laptops');
    });

    test('falls back to the single name when a language has none', () {
      final tree = CategoryTree([
        _c('a', null, en: 'Old'),
        Category(
          id: 'b',
          name: 'Only English',
          description: '',
          iconName: '',
          isActive: true,
          createdAt: _created,

          parentId: 'a',
        ),
      ]);
      expect(tree.pathLabel('b', 'ar'), 'Old › Only English');
    });

    test('the subtree of a category is itself and everything below it', () {
      final tree = CategoryTree(sample);
      expect(tree.subtreeIds('net'), {'net', 'rou', 'wifi', 'swi'});
      expect(tree.subtreeIds('rou'), {'rou', 'wifi'});
      expect(tree.subtreeIds('wifi'), {'wifi'});
      expect(tree.subtreeIds('ghost'), isEmpty);
    });

    test('has no fixed depth: 3,000 nested levels work', () {
      final chain = [
        for (var i = 0; i < 3000; i++) _c('c$i', i == 0 ? null : 'c${i - 1}'),
      ];
      final tree = CategoryTree(chain);
      expect(tree.subtreeIds('c0'), hasLength(3000));
      expect(tree.pathOf('c2999'), hasLength(3000));
      expect(tree.isEffectivelyActive('c2999'), isTrue);
    });

    test('never loops on a cycle in bad data', () {
      final tree = CategoryTree([_c('a', 'b'), _c('b', 'a'), _c('c', 'c')]);
      expect(tree.pathOf('a').map((c) => c.id).toSet(), {'a', 'b'});
      expect(tree.subtreeIds('a'), {'a', 'b'});
      expect(tree.pathOf('c').map((c) => c.id), ['c']);
    });

    test('a category whose parent is missing is treated as top-level', () {
      final tree = CategoryTree([_c('x', 'gone')]);
      expect(tree.childrenOf(null).map((c) => c.id), ['x']);
    });

    test('a category is only shown when it and every ancestor is active', () {
      final tree = CategoryTree([
        _c('r', null),
        _c('a', 'r', isActive: false),
        _c('b', 'a'),
        _c('c', 'r'),
        _c('d', 'r', deletionPending: true),
      ]);
      expect(tree.isEffectivelyActive('r'), isTrue);
      expect(tree.isEffectivelyActive('a'), isFalse);
      expect(tree.isEffectivelyActive('b'), isFalse); // hidden with its parent
      expect(tree.isEffectivelyActive('d'), isFalse); // being deleted
      expect(tree.isEffectivelyActive('ghost'), isFalse);
      expect(tree.activeChildrenOf('r').map((c) => c.id), ['c']);
    });

    test(
      'searches Arabic and English names, forgivingly, among shown ones',
      () {
        final tree = CategoryTree([
          ...sample,
          _c('hid', 'net', en: 'Hidden Wi-Fi', isActive: false),
        ]);
        expect(tree.search('wi-fi').map((c) => c.id), ['wifi']);
        expect(tree.search('واي فاي').map((c) => c.id), ['wifi']);
        expect(tree.search('راوتر').map((c) => c.id), ['rou']);
        expect(tree.search('لابتوب').map((c) => c.id), ['lap']);
        expect(tree.search('   '), isEmpty);
        expect(tree.search('zzz'), isEmpty);
      },
    );
  });

  group('the category data model', () {
    test('reads the tree fields', () {
      final category = CategoryModel.fromMap('k', {
        'name': 'Routers',
        'nameAr': 'راوترات',
        'nameEn': 'Routers',
        'parentId': 'net',
        'ancestorIds': ['top', 'net'],
        'deletionPending': true,
        'isActive': true,
      });
      expect(category.parentId, 'net');
      expect(category.ancestorIds, ['top', 'net']);
      expect(category.deletionPending, isTrue);
    });

    test('an older category has no parent or marker', () {
      final category = CategoryModel.fromMap('k', {
        'name': 'Old',
        'isActive': true,
      });
      expect(category.parentId, isNull);
      expect(category.ancestorIds, isEmpty);
      expect(category.deletionPending, isFalse);
    });

    test('bad values are ignored, never a crash', () {
      final category = CategoryModel.fromMap('k', {
        'name': 'x',
        'parentId': '',
        'ancestorIds': ['a', 5, null, 'b'],
        'deletionPending': 'yes',
      });
      expect(category.parentId, isNull);
      expect(category.ancestorIds, ['a', 'b']);
      expect(category.deletionPending, isFalse);
    });
  });

  group('changes made by Platform Admin appear without any app update', () {
    test(
      'a new category shows up in the tree as soon as it is saved',
      () async {
        final feed = StreamController<List<Category>>();
        final container = ProviderContainer(
          overrides: [allCategoriesProvider.overrideWith((ref) => feed.stream)],
        );
        addTearDown(() async {
          container.dispose();
          await feed.close();
        });
        final seen = <List<String>>[];
        container.listen(
          categoryTreeProvider,
          (_, tree) => seen.add(tree.all.map((c) => c.id).toList()),
          fireImmediately: true,
        );
        expect(seen.last, isEmpty);

        feed.add([_c('p1', null)]);
        await pumpEventQueue();
        expect(seen.last, ['p1']);

        // Platform Admin adds a sub-category and another top-level category
        feed.add([_c('p1', null), _c('p2', 'p1'), _c('s1', null)]);
        await pumpEventQueue();
        expect(seen.last.toSet(), {'p1', 'p2', 's1'});
        expect(
          container
              .read(categoryTreeProvider)
              .childrenOf('p1')
              .map((c) => c.id),
          ['p2'],
        );

        // ... and later deactivates the parent: the sub-category goes with it
        feed.add([
          _c('p1', null, isActive: false),
          _c('p2', 'p1'),
          _c('s1', null),
        ]);
        await pumpEventQueue();
        final tree = container.read(categoryTreeProvider);
        expect(tree.isEffectivelyActive('p2'), isFalse);
        expect(tree.activeChildrenOf(null).map((c) => c.id), ['s1']);
      },
    );

    test('paths are worded in the language the app shows', () async {
      final container = ProviderContainer(
        overrides: [
          allCategoriesProvider.overrideWith(
            (ref) => Stream.value([
              _c('a', null, ar: 'شبكات', en: 'Networking'),
              _c('b', 'a', ar: 'راوترات', en: 'Routers'),
            ]),
          ),
        ],
      );
      addTearDown(container.dispose);
      container.listen(allCategoriesProvider, (_, _) {});
      await pumpEventQueue();
      expect(
        container.read(categoryPathNamesProvider)['b'],
        'Networking › Routers',
      );
    });
  });
}

File: platform_admin_web/src/data/categoryTree.test.ts
import { describe, expect, it } from 'vitest';
import {
  buildIndex,
  chainForChildOf,
  childrenOf,
  compareSiblings,
  depthOf,
  descendantIds,
  findChainMismatches,
  interruptedDeletions,
  isEffectivelyActive,
  moveProblem,
  nextSortOrderAmong,
  pathLabel,
  pathOf,
  subtreeIds,
  visibleRows,
} from './categoryTree';
import type { Category } from './types';

const category = (id: string, parentId: string | null, extra: Partial<Category> = {}): Category => ({
  id,
  name: id,
  nameAr: '',
  nameEn: '',
  parentId,
  ancestorIds: [],
  deletionPending: false,
  sortOrder: null,
  description: '',
  iconName: '',
  isActive: true,
  createdAt: null,
  ...extra,
});

/** Builds categories from [id, parentId] pairs, parents first, with correct chains. */
function tree(edges: [string, string | null][], extra: Partial<Category> = {}) {
  const chains = new Map<string, string[]>();
  return edges.map(([id, parent]) => {
    const chain = parent ? [...(chains.get(parent) ?? []), parent] : [];
    chains.set(id, chain);
    return category(id, parent, { ancestorIds: chain, ...extra });
  });
}

const sample = tree([
  ['r', null], ['a', 'r'], ['b', 'r'], ['a1', 'a'], ['a2', 'a'], ['a1a', 'a1'], ['z', null],
]);

describe('walking a tree', () => {
  const index = buildIndex(sample);

  it('orders siblings by position, then name', () => {
    const list = [
      category('b', null, { sortOrder: 1 }),
      category('a', null, { sortOrder: 0 }),
      category('c', null),
      category('d', null, { name: 'Alpha' }),
    ];
    expect([...list].sort(compareSiblings).map((c) => c.id)).toEqual(['a', 'b', 'd', 'c']);
  });

  it('lists children and every descendant at any depth', () => {
    expect(childrenOf(index, 'r').map((c) => c.id)).toEqual(['a', 'b']);
    expect(childrenOf(index, null).map((c) => c.id)).toEqual(['r', 'z']);
    expect(descendantIds(index, 'r').sort()).toEqual(['a', 'a1', 'a1a', 'a2', 'b']);
    expect(subtreeIds(index, 'a1')).toEqual(['a1', 'a1a']);
    expect(descendantIds(index, 'z')).toEqual([]);
  });

  it('gives the path from the top and its depth', () => {
    expect(pathOf(index, 'a1a').map((c) => c.id)).toEqual(['r', 'a', 'a1', 'a1a']);
    expect(depthOf(index, 'a1a')).toBe(3);
    expect(depthOf(index, 'r')).toBe(0);
  });

  it('shows the whole path in the language shown, falling back to the single name', () => {
    const named = buildIndex([
      category('n', null, { name: 'Networking', nameAr: 'شبكات', nameEn: 'Networking' }),
      category('r', 'n', { name: 'Routers', nameAr: 'راوترات', nameEn: 'Routers' }),
      category('w', 'r', { name: 'Wi-Fi 6' }),
    ]);
    expect(pathLabel(named, 'w', 'en')).toBe('Networking › Routers › Wi-Fi 6');
    expect(pathLabel(named, 'w', 'ar')).toBe('شبكات › راوترات › Wi-Fi 6');
    expect(pathLabel(named, 'w', 'en', ' / ')).toBe('Networking / Routers / Wi-Fi 6');
  });

  it('handles a chain of 2,000 levels without a fixed limit or a stack overflow', () => {
    const edges: [string, string | null][] = [];
    for (let i = 0; i < 2000; i++) edges.push([`c${i}`, i === 0 ? null : `c${i - 1}`]);
    const deep = buildIndex(tree(edges));
    expect(descendantIds(deep, 'c0')).toHaveLength(1999);
    expect(pathOf(deep, 'c1999')).toHaveLength(2000);
    expect(chainForChildOf(deep, 'c1999')).toHaveLength(2000);
  });

  it('never loops on bad data with a cycle', () => {
    const broken = buildIndex([category('a', 'b'), category('b', 'a'), category('c', 'c')]);
    expect(descendantIds(broken, 'a').sort()).toEqual(['b']);
    expect(pathOf(broken, 'a').map((c) => c.id).sort()).toEqual(['a', 'b']);
    expect(pathOf(broken, 'c').map((c) => c.id)).toEqual(['c']);
  });

  it('treats a category whose parent is missing as top-level, so nothing disappears', () => {
    const orphaned = buildIndex([category('x', 'gone')]);
    expect(childrenOf(orphaned, null).map((c) => c.id)).toEqual(['x']);
  });
});

describe('moving is only allowed where it cannot make a cycle', () => {
  const index = buildIndex(sample);

  it('refuses itself, its own descendants, no change and missing categories', () => {
    expect(moveProblem(index, 'a', 'a')).toBe('self');
    expect(moveProblem(index, 'a', 'a1')).toBe('descendant');
    expect(moveProblem(index, 'r', 'a1a')).toBe('descendant');
    expect(moveProblem(index, 'a', 'r')).toBe('same');
    expect(moveProblem(index, 'r', null)).toBe('same');
    expect(moveProblem(index, 'a', 'ghost')).toBe('missing');
    expect(moveProblem(index, 'ghost', 'r')).toBe('missing');
  });

  it('allows a sibling, a cousin, an ancestor and the top level', () => {
    expect(moveProblem(index, 'a1', 'b')).toBeNull();
    expect(moveProblem(index, 'a1a', 'z')).toBeNull();
    expect(moveProblem(index, 'a1a', 'r')).toBeNull();
    expect(moveProblem(index, 'a1', null)).toBeNull();
  });

  it('refuses a parent that is awaiting deletion', () => {
    const marked = buildIndex([...sample, category('dying', null, { deletionPending: true })]);
    expect(moveProblem(marked, 'a', 'dying')).toBe('pending');
  });
});

describe('keeping the stored chains right', () => {
  it('a consistent tree has no mismatch', () => {
    expect(findChainMismatches(sample)).toEqual([]);
  });

  it('finds the categories whose chain does not match their parent links', () => {
    const wrong = sample.map((c) => (c.id === 'a1a' ? { ...c, ancestorIds: ['r'] } : c));
    expect(findChainMismatches(wrong).map((c) => c.id)).toEqual(['a1a']);
    const moved = sample.map((c) => (c.id === 'a' ? { ...c, parentId: 'z' } : c));
    // a's own chain and all its descendants' chains are now stale
    expect(findChainMismatches(moved).map((c) => c.id).sort()).toEqual(['a', 'a1', 'a1a', 'a2']);
  });

  it('gives the chain a child of a category must store', () => {
    const index = buildIndex(sample);
    expect(chainForChildOf(index, null)).toEqual([]);
    expect(chainForChildOf(index, 'a1')).toEqual(['r', 'a', 'a1']);
  });

  it('numbers a new category after its siblings', () => {
    const list = tree([['r', null], ['a', 'r'], ['b', 'r']]).map((c, i) => ({
      ...c,
      sortOrder: i === 2 ? 7 : null,
    }));
    const index = buildIndex(list);
    expect(nextSortOrderAmong(index, 'r')).toBe(8);
    expect(nextSortOrderAmong(index, 'a')).toBe(0);
  });
});

describe('what customers and companies see', () => {
  it('a category is only shown when it and every ancestor is active', () => {
    const list = tree([['r', null], ['a', 'r'], ['b', 'a']]).map((c) =>
      c.id === 'a' ? { ...c, isActive: false } : c,
    );
    const index = buildIndex(list);
    expect(isEffectivelyActive(index, 'r')).toBe(true);
    expect(isEffectivelyActive(index, 'a')).toBe(false);
    expect(isEffectivelyActive(index, 'b')).toBe(false); // hidden with its parent
    expect(isEffectivelyActive(index, 'ghost')).toBe(false);
  });

  it('a category awaiting deletion is not shown', () => {
    const index = buildIndex([category('x', null, { deletionPending: true })]);
    expect(isEffectivelyActive(index, 'x')).toBe(false);
  });

  it('finds the top-most category of each interrupted deletion', () => {
    const list = tree([['r', null], ['a', 'r'], ['b', 'a'], ['s', null]]).map((c) =>
      ['r', 'a', 'b'].includes(c.id) ? { ...c, deletionPending: true } : c,
    );
    expect(interruptedDeletions(list).map((c) => c.id)).toEqual(['r']);
  });
});

describe('the rows of the tree page', () => {
  const list = tree([['r', null], ['a', 'r'], ['a1', 'a']]).map((c) => ({
    ...c,
    nameAr: c.id === 'r' ? 'شبكات' : c.id === 'a' ? 'راوترات' : 'واي فاي',
    nameEn: c.id === 'r' ? 'Networking' : c.id === 'a' ? 'Routers' : 'Wi-Fi 6',
  }));

  it('shows only top-level rows until a branch is expanded', () => {
    const rows = visibleRows(list, new Set(), '');
    expect(rows.map((r) => [r.category.id, r.depth, r.hasChildren, r.expanded])).toEqual([
      ['r', 0, true, false],
    ]);
  });

  it('shows expanded branches at their depth', () => {
    const rows = visibleRows(list, new Set(['r', 'a']), '');
    expect(rows.map((r) => [r.category.id, r.depth])).toEqual([['r', 0], ['a', 1], ['a1', 2]]);
  });

  it('searches by Arabic or English name and opens the way to a match', () => {
    const en = visibleRows(list, new Set(), 'wi-fi');
    expect(en.map((r) => [r.category.id, r.matched, r.expanded])).toEqual([
      ['r', false, true],
      ['a', false, true],
      ['a1', true, false],
    ]);
    const ar = visibleRows(list, new Set(), 'واي فاي');
    expect(ar.map((r) => r.category.id)).toEqual(['r', 'a', 'a1']);
    expect(visibleRows(list, new Set(), 'راوترات').map((r) => r.category.id)).toEqual(['r', 'a']);
    expect(visibleRows(list, new Set(), 'nothing here')).toEqual([]);
  });

  it('lists older categories (no parent, no chain) as top-level ones', () => {
    const withOld = [...list, category('old', null)];
    expect(visibleRows(withOld, new Set(), '').map((r) => r.category.id)).toEqual(['old', 'r']);
    expect(findChainMismatches(withOld)).toEqual([]);
  });
});
